### PR TITLE
[supervisor] shutdown processes gracefully

### LINKED ISSUE #13365: [supervisor] Processes started in terminals do not get a SIGTERM signal on stop
We only SIGTERM the main process of the terminal and rely on propagation which doesn't seem to happen with the interactive bash processes.

We need to make sure we SIGTERM all the processes started within the terminals before SIGTERMING the main process.

### PR DESCRIPTION
**Pull request by svenefftinge**
## Description
Changes the shutdown logic so that on SIGTERM or SIGINT on the `supervisor init` process we are sending SIGTERM to `supervisor` which in turn sends SIGTERM on all its terminal processes.
When supervisor finished we iterate over all of `init`'s child processes until the grace period is up.

## Related Issue(s)

Fixes #13365

## How to test
Start a workspace https://se-supervi88c18dbe3b.preview.gitpod-dev.com/#https://github.com/svenefftinge/termination-test

Run 
```
./ignoresigterm > out_disown.txt &
disown $1
./ignoresigterm > out_bg.txt &
./ignoresigterm > out.txt
```
and stop the workspace.
With this preview environment you should see log output like below in all three files
```
awaiting signal

Received signal
terminated

Ignored for 0 seconds.
Ignored for 1 seconds.
Ignored for 2 seconds.
Ignored for 3 seconds.
Ignored for 4 seconds.
Ignored for 5 seconds.
Ignored for 6 seconds.
Ignored for 7 seconds.
Ignored for 8 seconds.
Ignored for 9 seconds.
Ignored for 10 seconds.
Ignored for 11 seconds.
Ignored for 12 seconds.
Ignored for 13 seconds.
Ignored for 14 seconds.
Ignored for 15 seconds.
...
```

With gitpod.io you will see only (i.e. the processes never received SIGINT):
```
awaiting signal
```

## Release Notes

```release-note
All running processes receive a SIGTERM when a workspace shuts down.
```

## Werft options:

- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`

### CODE REVIEW

**Comment by roboquat** [comment]
@svenefftinge: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by svenefftinge** [comment]
Adding hold because I want to make sure this gets thoroughly tested.

/hold

**Comment by svenefftinge** [comment]
@csweichel One thing I noticed is that when I start the hanging go process directly in a terminal bash and then `kill -15 <PID>` the bash, the go process dies as well without a SIGTERM but is still listed as a zombie process in the proc table.
When I nest this in another bash and the SIGTERM the nested bash, the go process gets reparented and continues to run.

File: components/common-go/process/process.go
Comment (svenefftinge) [inline components/common-go/process/process.go:30]: bad API design. pass in a <-chan instead

File: components/supervisor/cmd/init.go
Comment (svenefftinge) [inline components/supervisor/cmd/init.go:109]: give go routines a but of time to send the sigkills.

File: components/supervisor/cmd/init.go
Comment (svenefftinge) [inline components/supervisor/cmd/init.go:66]: move up

File: components/supervisor/pkg/supervisor/config.go
Comment (svenefftinge) [inline components/supervisor/pkg/supervisor/config.go:416]: todo validation

File: components/common-go/process/process.go
Comment (easyCZ) [inline components/common-go/process/process.go:29]: Why do you use generics here?

I think if all you're looking for is a signal from the outside then the idiomatic would be `chan struct{}`

File: components/supervisor/cmd/init.go
Comment (easyCZ) [inline components/supervisor/cmd/init.go:71]:
```suggestion
			terminationDone := make(chan struct{})
```
is more idiomatic and uses fewer bytes

File: components/common-go/process/process.go
Comment (easyCZ) [inline components/common-go/process/process.go:40]: Normally I'd see this written as 
```
if err != nil {
  if err == os.ErrProcessDone { ..}
  otherwise..
}
```
but I don't think it matters much

File: components/common-go/process/process.go
Comment (easyCZ) [inline components/common-go/process/process.go:29]: Is it acceptable to go terminate the process even if at the start of this method the channel already has a value?

If not, you can do
```
select {
  case <- gracePeriodOver:
      return ..
   default:
}
```
Which will check the channel, but if no value is available it will continue

File: components/common-go/process/process.go
Comment (easyCZ) [inline components/common-go/process/process.go:44]: If the process never stops, and we reach `gracePeriodOver` we will leak this go-routine and also the `terminated` channel.

What's the expected behaviour here? Is it that `process.Kill()` causes `process.Wait()` to unblock?

File: components/supervisor/cmd/init.go
Comment (easyCZ) [inline components/supervisor/cmd/init.go:68]: For readability, I'd move this whole block into a separate function

File: components/supervisor/cmd/init.go
Comment (easyCZ) [inline components/supervisor/cmd/init.go:111]: This would leak the goroutine (and terminationDone channel) above. Does that matter?

**Review by easyCZ — COMMENTED** [review]
Left a couple of Qs.

In general, I feel a lot of the code would become much more readable if we synchronised using a `context.Context`, in particular using the context.Done() property ([docs](https://pkg.go.dev/context#WithCancel)) than passing a channel as input to a function.

Also beware of the specific of `time.After(x)`. If the value is not read, it leaks - this can happen here when the processes do terminate within the period gracefully. See https://www.arangodb.com/2020/09/a-story-of-a-memory-leak-in-go-how-to-properly-use-time-after/

**Comment by svenefftinge** [comment]
I decided against using context because of the side effect of kill the process when done. That seems surprising.

**Comment by easyCZ** [comment]
> I decided against using context because of the side effect of kill the process when done. That seems surprising.

I'm not sure I follow. Could you explain, please?

I'd have expected that at the point we begin to handle the termination, we do the following:
```
ctx, cancel := context.WithDeadline(context.Background(), period)

// pass context into relevant methods, each async method does:
select {
  case <-ctx.Done():
	 exit..
  case <- something else
 
}
```

File: components/common-go/process/process.go
Comment (svenefftinge) [reply components/common-go/process/process.go:44]: I haven't thought of the leakage of the routine, but, yes, process.Kill() will unblock Wait().

File: components/supervisor/cmd/init.go
Comment (svenefftinge) [reply components/supervisor/cmd/init.go:111]: The process ends here.

**Comment by svenefftinge** [comment]
We don't exit but force kill the process, which I find a bit surprising if this was a side effect of a context cancellation:
```go
case <-gracePeriodOver:
		err = process.Kill()
		if err != nil {
			return err
		}
		return ErrForceKilled
	}
```

File: components/supervisor/cmd/init.go
Comment (easyCZ) [reply components/supervisor/cmd/init.go:111]: In that case, should it contain `os.Exit(code)` then? It would at least communicate if it terminated gracefully, or not.

**Review by easyCZ — COMMENTED** [review]
How would you feel about at least logging the PIDs being killed? I feel that for the amount of complexity this contains, we've got zero visibility into what it's actually doing.

File: components/supervisor/cmd/init.go
Comment (easyCZ) [reply components/supervisor/cmd/init.go:111]: Also, I think it's worth flushing the logger at this point to guarantee any unwritten bytes are actually written.

**Comment by easyCZ** [comment]
> which I find a bit surprising if this was a side effect of a context cancellation:

But the context doesn't kill the processes. It is just a signal that you've exceeded the grace period (or the time allowed for you to propagate the signal into) - in this respect, it's the same as a call stack of request. It's more less replacing the current argument you're passing in to indicate it's run out of time, just more standardized and wider used.

If you want, happy to pair to see what this would look like otherwise it matters not.

File: components/common-go/process/process.go
Comment (svenefftinge) [reply components/common-go/process/process.go:29]: It allows passing in `<- chan time.Time` (from time.After) but also other chans and communicate that we are only interested in the event.

**Comment by svenefftinge** [comment]
Ok, I think if we put it like that it's ok to use context. API-wise it fits better and takes care of the leakage you mentioned

File: components/common-go/process/process.go
Comment (svenefftinge) [reply components/common-go/process/process.go:29]: Thanks, we do want run through the code including the effect of SIGKILL at the end.

File: components/supervisor/cmd/init.go
Comment (svenefftinge) [reply components/supervisor/cmd/init.go:90]: > In that case, should it contain os.Exit(code) then? It would at least communicate if it terminated gracefully, or not.

But it just ends regularly here. So no reason to exit with an error code. Do I miss something?

Comment (easyCZ) [reply components/supervisor/cmd/init.go:90]: Normally, when you do not terminate gracefully you signal that by setting a non-zero exit code to the shell. So in this case, we'd set it when the deadline is reached

Comment (svenefftinge) [reply components/supervisor/cmd/init.go:90]: But it is not this process that doesn't terminate gracefully, but processes that were started by users during the workspace session. I don't think we should signal through error code on this level if such a process doesn't behave well

Comment (easyCZ) [reply components/supervisor/cmd/init.go:90]: Good point. I guess what I'm after is to be able to tell the user (or even measure it ourselves) when we exit without completing the termination - this would then signal to users that they should possibly increase the grace termination period (in their config) but in general help them understand if the stop task did complete or not.

Comment (svenefftinge) [reply components/supervisor/cmd/init.go:90]: Yes, that's important. The plan is to write the stdout and stderr of "shutdown" tasks [[pr](https://github.com/gitpod-io/gitpod/pull/11287)] to some file in `/workspace/.gitpod`. Do you think that is good enough? We could alternatively also write a shutdown protocol to that location.

Comment (svenefftinge) [reply components/supervisor/cmd/init.go:90]: I think I like the idea of writing a termination protocol to a file in that folder. Will look into it.

File: components/supervisor/cmd/init.go
Comment (akosyakov) [inline components/supervisor/cmd/init.go:106]: return is missing?

**Comment by akosyakov** [comment]
@svenefftinge everything but `./ignoresigterm > out.txt` works for me. I tried from both supervisor and VS Code terminals.

**Review by akosyakov — APPROVED** [review]
ok, my bad, it turned out that one need to trap SIGHUP otherwise foreground processes don't receive a signal

**Pull request merged**
